experiment/telegram/telegram.go

```diff
@@ -18,27 +22,101 @@ const (
 )
 
 // Config contains the experiment config.
-type Config struct {
-	// LogLevel is the MK log level. Empty implies "WARNING".
-	LogLevel string
+type Config struct{}
+
+// TestKeys contains telegram test keys.
+type TestKeys struct {
+	TelegramHTTPBlocking bool                           `json:"telegram_http_blocking"`
+	TelegramTCPBlocking  bool                           `json:"telegram_tcp_blocking"`
+	TelegramWebFailure   *string                        `json:"telegram_web_failure"`
+	TelegramWebStatus    *string                        `json:"telegram_web_status"`
+	TCPConnect           []ootemplate.TCPConnectResults `json:"tcp_connect"`
+	Requests             []ootemplate.HTTPRoundTrip     `json:"requests"`
 }
 
 func measure(
 	ctx context.Context, sess *session.Session, measurement *model.Measurement,
 	callbacks handler.Callbacks, config Config,
 ) error {
-	settings := measurementkit.NewSettings(
-		"Telegram", sess.SoftwareName, sess.SoftwareVersion,
-		sess.CABundlePath(), sess.ProbeASNString(), sess.ProbeCC(),
-		sess.ProbeIP(), sess.ProbeNetworkName(), config.LogLevel,
-	)
-	out, err := measurementkit.StartEx(settings, sess.Logger)
+	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
+	testkeys := &TestKeys{
+		TelegramHTTPBlocking: true,
+		TelegramTCPBlocking:  true,
+	}
+	measurement.TestKeys = testkeys
+	defer func() {
+		for _, tcpconn := range testkeys.TCPConnect {
+			if tcpconn.Status.Success == true {
+				testkeys.TelegramTCPBlocking = false
+				break
+			}
+		}
+		for _, roundtrip := range testkeys.Requests {
+			if strings.Contains(roundtrip.Request.URL, "web.telegram.org") {
+				continue
+			}
+			if roundtrip.Failure == "" {
+				testkeys.TelegramHTTPBlocking = false
+				break
+			}
+		}
+		var failure *string
+		count := 0
+		for _, roundtrip := range testkeys.Requests {
+			if !strings.Contains(roundtrip.Request.URL, "web.telegram.org") {
+				continue
```

Comment:
Why is this check being done to ensure the URL contains the string web.telegram.org?

Reply:
Because the HTTP results include both web and non-web endpoints and they are used to compute different sets of output variables. Hence, in this specific case, I am making sure I am not using web endpoints because they should not be used to compute this stat.

Reply:
It's probably a good idea to mention this in a comment, it was not immediate to me that was the reason.

Reply:
Yeah, I did that here: https://github.com/ooni/probe-engine/pull/76/files#diff-57ae86adcc9a6f48b8959ad307776b49R91